### PR TITLE
feat: Upgrade to V3

### PR DESCRIPTION
**Pull request by vediyappan-hcl**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [Y] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [N] I am not introducing a new dependency (add notes below if you are)
- [N] I have added unit tests for the new feature or bug fix (if not, why?)
- [Y] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [N] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions
1. Run the device-uart service along with edgex-go services. Before that, ensure you're mentioning correct deviceName (i.e., /dev/ttyUSB0) and baudRate (i.e., 115200) in device file and device profile.
2. Read data send from a serial device (I used Docklight serial application) using core-command:
`curl -X GET http://localhost:59882/api/v3/device/name/Uart-Generic-Device/Read_UART | json_pp`
3. Write data to a serial device some data using core-command:
`curl -X PUT -d '{"Write_UART":"57656c636f6d6520746f2074686520456d62656464656420576f726c6421"}'  http://localhost:59882/api/v3/device/name/Uart-Generic-Device/Write_UART | json_pp`
4. Cross verify the device-uart logs (optional).
```
level=INFO ts=2023-06-27T07:50:29.610027287Z app=device-uart source=driver.go:100 msg="Driver.HandleReadCommands(): Received Data =  456d62656464656420506c6174666f726d204c6162"
level=INFO ts=2023-06-27T07:50:06.116906919Z app=device-uart source=driver.go:122 msg="Driver.HandleWriteCommands(): protocols: map[UART:map[baudRate:115200 deviceName:/dev/ttyAMA2]], resource: Write_UART, attribute: map[baud:115200 dev:/dev/ttyAMA2 timeout:1 type:generic], parameters: [DeviceResource: Write_UART, String: 57656c636f6d6520746f2074686520456d62656464656420576f726c6421]"
```

## New Dependency Instructions
+    "github.com/gorilla/mux"
+    "github.com/spf13/cast"

## Additional Info

- I will open a PR for the related docs change after discussion with Lindsey Cheng.
- Attached test logs below:

[Logs_Device-UART-v3_Read-and-Write_Success.txt](https://github.com/edgexfoundry/device-uart/files/11880261/Logs_Device-UART-v3_Read-and-Write_Success.txt)
Docklight_Device-UART-v3_TX-RX-Success

### CODE REVIEW

**Comment by lindseysimple** [comment]
@lenny-intel Lenny would you please help review this PR when you have time? This seems to cover https://github.com/edgexfoundry/device-uart/issues/12 you opened earlier, thanks!

**Comment by lenny-intel** [comment]
Hi @vediyappan-hcl, please squash your commits and "sign-off" on the single commit message using `-s` flag. 
Also the message and PR title must use Conventional Commit syntax.
See https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md for details.
This will clear up the above failed checks.

File: cmd/res/configuration.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# REFERENCE: https://docs.edgexfoundry.org/3.0/microservices/device/Ch-DeviceServices/
+Writable:
+  LogLevel: INFO
+  Reading:
+    ReadingUnits: true
+  Telemetry:
+    Metrics:
+      EventsSent: false
+      ReadingsSent: false
```
Comment (lenny-intel) [inline cmd/res/configuration.yaml:9]: Remove, these are in the new common config here: https://github.com/edgexfoundry/edgex-go/blob/main/cmd/core-common-config-bootstrapper/res/configuration.yaml#L106-L111

File: cmd/res/configuration.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  Port: 59995
+  StartupMsg: device uart started
+  RequestTimeout: "5s"
+
+MessageBus:
+  Protocol: redis
+  Host: localhost # uncomment when running from command-line in hybrid mode
+  Port: 6379
+  Type: redis
+  BaseTopicPrefix: "edgex" # prepended to all topics as "edgex/<additional topic levels>
```
Comment (lenny-intel) [inline cmd/res/configuration.yaml:22]: Remove, these are in the new common config here: https://github.com/edgexfoundry/edgex-go/blob/main/cmd/core-common-config-bootstrapper/res/configuration.yaml#L54-L77

File: cmd/res/configuration.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+Clients:
+  core-metadata:
+    Protocol: "http"
+    Host: "localhost"
+    Port: 59881
+
+Device:
+  DataTransform: true
+  MaxCmdOps: 128
+  MaxCmdValueLen: 256
```
Comment (lenny-intel) [inline cmd/res/configuration.yaml:35]: Remove, these are in the new common config here: https://github.com/edgexfoundry/edgex-go/blob/main/cmd/core-common-config-bootstrapper/res/configuration.yaml#L117C8-L130C22

File: cmd/res/configuration.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  ProfilesDir: "./res/profiles"
+  DevicesDir: "./res/devices"
+  # ProvisionWatchersDir is omitted here since most Device Services don't use it.
+  # Those that do will have it in their private config
+  EnableAsyncReadings: true
+  AsyncBufferSize: 16
+  Labels: []
+  Discovery:
+    Enabled: false
+    Interval: "30s"
```
Comment (lenny-intel) [inline cmd/res/configuration.yaml:45]: Remove, these are in the new common config here: https://github.com/edgexfoundry/edgex-go/blob/main/cmd/core-common-config-bootstrapper/res/configuration.yaml#L117C8-L130C22

File: cmd/res/configuration.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    Protocol: "http"
+    Host: "localhost"
+    Port: 59881
+
+Device:
+  DataTransform: true
+  MaxCmdOps: 128
+  MaxCmdValueLen: 256
+  ProfilesDir: "./res/profiles"
+  DevicesDir: "./res/devices"
```
Comment (lenny-intel) [inline cmd/res/configuration.yaml:37]: Note that we leave these so the can be overridden with Env Vars, even though they are in the common config.
See here for example: https://github.com/edgexfoundry/device-virtual-go/blob/main/cmd/res/configuration.yaml#L23-L26

File: cmd/res/devices/devices.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# Pre-defined devices
```
Comment (lenny-intel) [inline cmd/res/devices/devices.yaml:1]: The TOML version needs to be removed.
https://github.com/edgexfoundry/device-uart/blob/main/cmd/res/devices/device.uart.toml

File: cmd/res/profiles/profiles.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+name: "Uart-Generic-Device"
```
Comment (lenny-intel) [inline cmd/res/profiles/profiles.yaml:1]: was this file renamed or copied?
Looks identical to https://github.com/edgexfoundry/device-uart/blob/main/cmd/res/profiles/device.uart.generic.yaml
file name should be `generic.yaml`. `profiles.yaml` name is ambiguous and and can only contain a single profile.

File: internal/driver/drv_uart_generic.go [hunk 8]
```diff
@@ -57,6 +82,323 @@ func NewUartGeneric(dev string, baud int, timeout int) *UartGeneric {
...
+        case common.ValueTypeFloat64:
+                valFloat := val.(float64)
+                if math.Abs(valFloat) >= math.SmallestNonzeroFloat64 && math.Abs(valFloat) <= math.MaxFloat64 {
+                        return nil
+                }
+        }
+
+        return fmt.Errorf("value %v for %s type is out of range", val, valueType)
+}
+
```
Comment (lenny-intel) [inline internal/driver/drv_uart_generic.go:401]: Why is all this new code being added?

This is out of scope for migrating to 3.0, please remove.

File: cmd/res/configuration.yaml (hunk 1 above)
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:9]: Removed

File: cmd/res/configuration.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  Telemetry:
+    Metrics:
+      EventsSent: false
+      ReadingsSent: false
+
+Service:
+  Host: localhost
+  Port: 59995
+  StartupMsg: device uart started
+  RequestTimeout: "5s"
```
[the remark this answers is not in the document]
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:15]:
```
Creates the below error. Hence retaining it.

vediyappan@vediyappan-desktop:~/work/uart-upgrade/device-uart/cmd$ ./device-uart
level=INFO ts=2023-06-28T11:34:52.216014507Z app=device-uart source=config.go:640 msg="Loading configuration file from res/configuration.yaml"
level=INFO ts=2023-06-28T11:34:52.218913423Z app=device-uart source=config.go:226 msg="Private configuration loaded from file with 0 overrides applied"
**level=ERROR ts=2023-06-28T11:34:52.220629214Z app=device-uart source=httpserver.go:101 msg="unable to parse RequestTimeout value of  to a duration: time: invalid duration \"\""
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x72a9f0]**

goroutine 1 [running]:
github.com/edgexfoundry/device-sdk-go/v3/pkg/startup.Bootstrap({0x90ca11, 0xb}, {0xa2b6d4, 0x5}, {0xa3b890?, 0x40003f9a80})
        /home/vediyappan/go/pkg/mod/github.com/edgexfoundry/device-sdk-go/v3@v3.0.0/pkg/startup/bootstrap.go:27 +0x2a0
main.main()
        /home/vediyappan/work/uart-upgrade/device-uart/cmd/main.go:23 +0x50
vediyappan@vediyappan-desktop:~/work/uart-upgrade/device-uart/cmd$
```

File: cmd/res/configuration.yaml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  Type: redis
+  BaseTopicPrefix: "edgex" # prepended to all topics as "edgex/<additional topic levels>  
+  Optional:
+    ClientId: device-uart
+
+Clients:
+  core-metadata:
+    Protocol: "http"
+    Host: "localhost"
+    Port: 59881
```
[the remark this answers is not in the document]
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:30]: Removing "core-metadata" clients configuration creates the below error. Hence retaining it.

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x694d28]

Note: This device-uart service runs with other EdgeX Core Services, such as Core Metadata, Core Data, and Core Command
Reference Link: https://github.com/edgexfoundry/device-uart#usage

File: cmd/res/configuration.yaml (hunk 5 above)
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:37]: Retained.

File: cmd/res/configuration.yaml (hunk 4 above)
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:45]: Removed.

File: cmd/res/devices/devices.yaml (hunk 6 above)
Comment (vediyappan-hcl) [reply cmd/res/devices/devices.yaml:1]: Removed res/devices/device.uart.toml.

File: cmd/res/profiles/profiles.yaml (hunk 7 above)
Comment (vediyappan-hcl) [reply cmd/res/profiles/profiles.yaml]: Renamed res/profiles/device.uart.generic.yaml to res/profiles/generic.yaml

File: internal/driver/driver.go [hunk 11]
```diff
@@ -23,34 +23,48 @@ package driver
...
 
     return nil
 }
 
+// Start runs device service startup tasks after the SDK has been completely
+// initialized. This allows device service to safely use DeviceServiceSDK
+// interface features in this function call
+func (s *Driver) Start() error {
+    handler := NewUartHandler(s.sdk)
+    return handler.Start()
```
[the remark this answers is not in the document]
Comment (vediyappan-hcl) [reply internal/driver/driver.go:65]: Addressed.

File: internal/driver/drv_uart_generic.go (hunk 8 above)
Comment (vediyappan-hcl) [reply internal/driver/drv_uart_generic.go:401]: Removed. 

We were having plan to support multiple datatypes i.e., Read and write integers. Hence, I had a thought of having it here. However, removed it as it is not in v3 migration scope.

File: cmd/res/configuration.yaml (hunk 2 above)
Comment (lenny-intel) [reply cmd/res/configuration.yaml:15]: See previous comment

File: cmd/res/configuration.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply cmd/res/configuration.yaml:23]: See previous comment

File: cmd/res/configuration.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:8]: Resolved.

Tested - Used configuration provider (consul), seeded common configuration by running core-common-config-bootstrapper service. Specified -cp -d flags when running the device service.

File: cmd/res/configuration.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:23]: Resolved.

Tested - Used configuration provider (consul), seeded common configuration by running core-common-config-bootstrapper service. Specified -cp -d flags when running the device service.

File: cmd/res/configuration.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+Device:
+  # These have common values (currently), but must be here for service local env overrides to apply when customized
+  ProfilesDir: "./res/profiles"
+  DevicesDir: "./res/devices"
+
+# value 0 unlimit the maximum event size that can be sent to message bus or core-data
+MaxEventSize: 0
```
[the remark this answers is not in the document]
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:31]: Resolved.

Tested - Used configuration provider (consul), seeded common configuration by running core-common-config-bootstrapper service. Specified -cp -d flags when running the device service.

File: cmd/res/configuration.yaml (hunk 2 above)
Comment (vediyappan-hcl) [reply cmd/res/configuration.yaml:15]: Resolved.

Tested - Used configuration provider (consul), seeded common configuration by running core-common-config-bootstrapper service. Specified -cp -d flags when running the device service.

**Comment by vediyappan-hcl** [comment]
@lenny-intel @lindseysimple Addressed the review comments. Could you please review it?

**Comment by vediyappan-hcl** [comment]
@lenny-intel @lindseysimple Addressed the review comments. Could you please review it?

**Review by lenny-intel — APPROVED** [review]
LGTM, but `gofmt` must be run on code first before we can merge. CI Pipeline is failing due to this.

**Comment by vediyappan-hcl** [comment]
> LGTM, but `gofmt` must be run on code first before we can merge. CI Pipeline is failing due to this.

@lenny-intel Addressed this one and all checks have passed now.

**Review by lenny-intel — APPROVED** [review]
LGTM
